File: COM1786-CW1-Project-5/COM1786-CW1-Project-5/app/src/main/java/com/example/comp1786cw1project3/feature/base/BaseFragment.java
package com.example.comp1786cw1project3.feature.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewbinding.ViewBinding;

import com.example.comp1786cw1project3.BaseActivity;
import com.tunjid.androidx.navigation.StackNavigator;

public abstract class BaseFragment<VB, VM> extends Fragment {
    private StackNavigator navigator;

    public VB viewBinding;
    protected abstract VM viewModel();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel();
        BaseActivity baseActivity = (BaseActivity) requireActivity();
        navigator = baseActivity.getStackNavigator();
    }

    abstract public VB onCreateViewBinding(LayoutInflater inflater);

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        viewBinding = onCreateViewBinding(inflater);
        ViewBinding a = (ViewBinding) viewBinding;
        a.getRoot();
        return a.getRoot();
    }

    public void navigate(Fragment fragment, Boolean clearAll) {
        if (clearAll) {
            navigator.clear(null, true);
        }
        navigator.push(fragment, fragment.getClass().getName());
    }

    public void navigateUp(String upToTag, Boolean includeMatch) {
        navigator.clear(upToTag, includeMatch);
    }

    public void navigateUp() {
        navigator.pop();
    }
}

File: COM1786-CW2-Project-5/COM1786-CW2-Project-5/app/src/main/java/com/example/comp1786cw2project2/feature/base/BaseFragment.kt
package com.example.comp1786cw2project2.feature.base

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModel
import androidx.viewbinding.ViewBinding
import com.example.comp1786cw2project2.MainActivity
import com.example.com1786_cw2project1.R
import com.tunjid.androidx.navigation.StackNavigator

abstract class BaseFragment<VB: ViewBinding, VM: ViewModel> : Fragment() {

    val stackNavigator: StackNavigator by lazy {
        (requireActivity() as MainActivity).stackNavigator
    }

    abstract val viewModel: VM?
    lateinit var viewBinding: VB

    abstract fun onCreateViewBinding(inflater: LayoutInflater): VB

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        viewBinding = onCreateViewBinding(inflater)
        return viewBinding.root
    }
}